### PR TITLE
[None][fix] Stabilize Gemma4 FA2 CUDA Graph decode on Hopper

### PR DESCRIPTION
**Pull request by lfr-0531**
## Description

Gemma4 uses FlashInfer FA2 attention on Hopper. Its H256/H512 CUDA Graph decode path needs tensor-core attention, while FlashInfer's split-K schedule depends on the runtime KV page-count distribution. Reusing a schedule for a different distribution can produce incorrect output, and disabling split-K avoids the issue at a significant performance cost.

This change keeps Gemma4 CUDA Graph decode and split-K enabled:

- Use tensor-core decode for high-head-dimension CUDA Graph plans.
- Re-plan the FA2 split-K schedule only when the per-request page-count distribution changes.
- Preserve the conservative CUDA Graph KV-length bound across runtime schedule refreshes.
- Keep FlashInfer's original pre-plan synchronization for safe pinned-workspace reuse.
- Sanitize evicted sliding-window page indices before FlashInfer can dereference them.

The implementation relies on FlashInfer 0.6.16's pinned CUDA Graph plan contract and does not duplicate its private `plan_info` layout or validation. No public API or dependency changes are included.

## Test Coverage

- Verified the installed Python sources match the tested source hashes and the native package targets SM90.
- H100 SM90 final-source unit validation:
  - Hopper split-K schedule-refresh regression: passed.
  - Six additional FA2 CUDA Graph tests covering E2B/26B/31B geometry, hybrid head dimensions, and multi-step decode: passed.
  - SWA raw `BAD_PAGE_INDEX` to sanitized metadata regression: passed.
- The split-K regression captures a skewed 12B-like H256/H512 distribution, replays a balanced distribution, then grows KV length within the same page; both replays match eager output.
- Reduced the regression KV allocation from 8192 to 512 blocks and verified that capacity is sufficient.
- Added the Gemma4 modeling filter to `l0_h100.yml`, consistent with the other model entries; Hopper FA2 coverage remains enabled while SM100-only `trtllm-gen` tests explicitly skip on H100.
- H200 SM90, Gemma4 E2B BF16 with CUDA Graph enabled:
  - GSM8K: 85.330, required >= 82.506.
  - MMLU: 56.750, required >= 55.006.
- H200 SM90, Gemma4 26B-A4B BF16 with CUDA Graph enabled:
  - GSM8K: 95.262, required >= 87.627.
- H100 real-model GSM8K graph/eager comparison on the same 20-sample subsets:
  - E2B: 75.0 / 75.0.
  - 12B: 90.0 / 90.0.
- H100 FA2 H512, KV length 4096, CUDA Graph microbenchmark (split-K vs disabled split-K):
  - 12B full-attention shape, batch 1: 0.01948 ms vs 0.34896 ms (17.92x).
  - 12B full-attention shape, batch 8: 0.04034 ms vs 0.35580 ms (8.82x).
  - E2B full-attention shape, batch 8: 0.03553 ms vs 0.34545 ms (9.72x).
  - Split-K and non-split outputs matched within BF16 tolerance.
- File-level pre-commit checks passed, including test-list AST validation.
- Python-only change; no C++ build was required.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Updated FlashInfer FA2 CUDA Graph decode planning for Gemma4.
- Preserved split-K planning and the conservative maximum KV-length bound.
- Refreshed schedules only when FA2 generation block-count distributions change.
- Restored required stream synchronization during decode replanning.
- Removed obsolete captured-plan metadata and host-side page-length tracking.
- Added `fa2_plan_num_blocks` to `FlashInferWrappers`.
- Added tensor-core graph selection before wrapper creation.
- Added SWA eviction-page sanitization coverage.
- Added the Gemma4 schedule-refresh test to the H100 pre-merge test list.
- No public API or dependency changes are reported.
- Review should verify CUDA Graph plan-contract assumptions and synchronization safety across all replanning paths.

## QA Engineer Review

- Modified Gemma4 CUDA Graph replay and split-K schedule-refresh tests.
- Reworked long-sequence VSWA generation coverage to validate evicted-page marker sanitization.
- Added `BAD_PAGE_INDEX` coverage.
- Removed redundant evicted-SWA CUDA Graph coverage.
- Added one Hopper-only 12B schedule-refresh test and retained split-K validation.
- Added the schedule-refresh test to `tests/integration/test_lists/test-db/l0_h100.yml`.
- Test-list coverage is present for the schedule-refresh test. Coverage for the other modified test functions is not confirmed from the provided summary.
- Verdict: needs follow-up.

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 1]
```diff
@@ -1323,7 +1340,18 @@ def _clean_cached_plans(self, *, defer_plan: bool):
...
+                wrappers = self._plan_params_to_wrappers[plan_params]
+                if wrappers.fa2_plan_num_blocks is not None:
+                    num_blocks = tuple(self.num_blocks[self.num_contexts:])
+                    if not num_blocks:
+                        wrappers.fa2_plan_num_blocks = None
+                    elif wrappers.fa2_plan_num_blocks == num_blocks:
+                        continue
+                    else:
+                        wrappers.is_planned = False
+                        self._plan_with_params(plan_params)
```
Comment (BowenFu) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py]: This branch calls `_plan_with_params()` unconditionally, but the caller in `prepare()` sets `defer_plan = len(active_wrappers) > 1` precisely because the Gemma4 hybrid case shares one `workspace_buffer` and "eager plan() would overwrite earlier wrappers' workspace". `fa2_plan_num_blocks` is only ever set on the `use_graph_tensor_cores` fa2 path, which for Gemma4 means both the H256 and H512 wrappers — so when the generation block count changes, both get planned eagerly here and the second `plan()` clobbers the first one's workspace. And since `_plan_with_params()` ends with `is_planned = True`, the deferred plan in `forward_impl` is then skipped, so nothing repairs it.

```suggestion
                        if not defer_plan:
                            self._plan_with_params(plan_params)
```

The re-plan still happens in `forward_impl` via `is_planned = False` on the line above, so the block-count-change path keeps working.

This one is required for this PR — it hits the multi-wrapper configuration the change targets.

Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py]: Thanks for flagging this. I rechecked the ownership and replay flow. Each FA2 decode wrapper owns a private integer plan workspace, and the shared float workspace is runtime scratch, so refreshing the H256 and H512 plans here does not overwrite their persistent schedules. The CUDA Graph replay path is `prepare(); graph.replay()` and does not re-enter Python `forward_impl`, so adding the proposed `if not defer_plan` guard would leave the changed split-K schedule stale. I kept the eager FA2 refresh, clarified this exception in the comments. I reran the exact Hopper schedule-refresh regression on H200, and it passed.

**Review by pengbowang-nv — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 2]
```diff
@@ -2042,7 +2091,14 @@ def decode_plan():
                 block_tables=block_tables,
                 # Keep FlashInfer's recorded graph shape aligned with the wrapper cache key.
                 q_len_per_req=plan_params.q_len_per_req,
+                disable_split_kv=False,
             )
+            if planned_max_kv_len is not None:
+                decode_wrapper._max_kv_len = max(planned_max_kv_len,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py]: Why we need to update `_max_kv_len`?

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 3]
```diff
@@ -1425,7 +1442,22 @@ def _clean_cached_plans(self, *, defer_plan: bool):
             # corresponding forward pass. So, flush them out here as they won't be relevant for
             # subsequent forward calls.
             if plan_params.attention_mask_data is None and plan_params.multi_item_params is None:
-                self._plan_params_to_wrappers[plan_params].is_planned = False
+                wrappers = self._plan_params_to_wrappers[plan_params]
+                if wrappers.fa2_plan_num_blocks is not None:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py:1446]: `_clean_cached_plans` is called in multiple places, it is hard to predict when will the `fa2_plan_num_blocks` re-plan be triggered. Can we move it to the end of `prepare`, so that we only trigger this kind of re-plan once after all operations to update num_blocks have happened?

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 2 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py]: Thanks for pointing this out. You are right that preserving `_max_kv_len` is unnecessary for this path. I removed the save/restore logic to avoid coupling to this private state.

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 3 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:1446]: Thanks, that makes sense. I moved the captured FA2 schedule refresh out of `_clean_cached_plans()` and now invoke it once near the end of `prepare()`, after the local page metadata updates are complete.
